### PR TITLE
stm32/uart: Add support for LPUART1 on L0, L4, H7 and WB MCUs.

### PR DESCRIPTION
**Pull request by chrismas9**
Add LPUART1 as a standard UART. No low power features are supported.
LPUART1 is enabled as the next available UART after the standard U(S)ARTs.
STM32WB: LPUART1 = UART(2)
STM32L0: LPUART1 = UART(6)
STM32L4: LPUART1 = UART(6)
STM32H7: LPUART1 = UART(9)

LPUART1 is enabled by defining MICROPY_HW_LPUART1_TX and MICROPY_HW_LPUART1_RX in mpconfigboard.h

### CODE REVIEW

**Comment by chrismas9** [comment]
Motivation for this PR was changing from F030 to L073 to get EEPROM. L073 does not have USART3, but has LPUART1 on same pin (PB10/11). It also adds a user UART to NUCLEO_WB55 which uses its only USART for REPL. Gives more pin choices for UART when normal UARTS unavailable due to other use of pins for I2C, ADC, Eth, etc.

Testing: LPUART1 as UART(6) tested with external loopbak and baud rate check on NUCLEO_L073RZ and NUCLEO_L452RE. Flow control not tested. I don't have hardware to test WB55 or H7. Checked UASART6 as UART(6) on PYBV3. NUCLEO_WB55 and NUCLEO_H743ZI ports compile with LPUART1 enabled. PYBV3 and NUCLEO_F767ZI compile without error. Review of all register differences between U(S)ART and LPUART. Code inspection of all direct access to STM32 registers. Handle difference in buadrate (BRR register). There are several interrupt enables in CR1 and CR2 not implemented in LPUART. These must be kept at reset state (0). They are defined, but only cleared, so have not changed the code.

**Comment by chrismas9** [comment]
I have another PR coming to reduce size of L073 port. Then I will update mpconfigport.h for various boards to enable the extra UARTs - LPUART1 and UASART4/5 on L0.

**Comment by dpgeorge** [comment]
Thanks for this.

> LPUART1 is enabled as the next available UART after the standard U(S)ARTs.

I'm not sure this is the right approach because it could get confusing with what the last normal UART number is (the STM32's have a big range in available UARTs).

I don't think it makes sense to have a `machine.LPUART` class because it's very stm32-specific.  Maybe something like `machine.UART('LP1')` to select the low-power peripherals?  Also need to think how the low-power functionality would be exposed, possibly just additional keyword args to `.init()`.

**Comment by chrismas9** [comment]
> I don't think it makes sense to have a machine.LPUART class

I agree, I didn't want to make any changes to the UART class, that's why I did it this way.
> Maybe something like machine.UART('LP1')

Good idea. Done
> Also need to think how the low-power functionality would be exposed, possibly just additional keyword args to .init().

I think the only low power feature that would be useful is to set the clock to 32kHz (9600 buad max) and enable wake on interrupt. I don't have a need of this and I don't know how popular it would be. I am not sure I would be confident implementing this.

File: ports/stm32/machine_uart.c [hunk 1]
```diff
@@ -147,7 +147,15 @@ STATIC const mp_irq_methods_t pyb_uart_irq_methods = {
 STATIC void pyb_uart_print(const mp_print_t *print, mp_obj_t self_in, mp_print_kind_t kind) {
     pyb_uart_obj_t *self = MP_OBJ_TO_PTR(self_in);
     if (!self->is_enabled) {
-        mp_printf(print, "UART(%u)", self->uart_id);
+        #ifdef LPUART1
+        if (self->uart_id == PYB_LPUART_1) {
+            mp_printf(print, "UART('MP1')");
```
Comment (H-Grobben) [inline ports/stm32/machine_uart.c:152]: Shouldn't this be LP1?

Comment (chrismas9) [reply ports/stm32/machine_uart.c:152]: I think it should be 'LP1", also line 182. I will review, rebase and retest this week

**Comment by H-Grobben** [comment]
I will merge this with my G4 setup ( #6808 ), and also with our custom WB board, and try testing later this week.

**Comment by dpgeorge** [comment]
I still think naming the UART with `"LP1"` and `"LP2"` etc is the right approach (rather than numbering them after all other UARTs).

**Comment by H-Grobben** [comment]
Tested succesfully it in our repo, only for G4 series Nucleo board:
https://github.com/AEMICS/micropython/commit/ba929ae7cace9c3021b9b7d4adfc1d7354faa4ad
(and some more commits for the checks for code formatting etc)
With this pull request, we don't need the additional wires on the Nucleo board!

**Comment by chrismas9** [comment]
> I still think naming the UART with "LP1" and "LP2" etc is the right approach (rather than numbering them after all other UARTs).

Yes, I agree now, especially if low power features are added later, in which case 'LP1', etc identify that these UARTs are different. The UARTs are enumerated in `uart.h` so the next number is used in the driver with 'LP1' handled as a special case. Implementing LPUART without using the next number would need more driver changes than I could do.

**Comment by chrismas9** [comment]
Rebased, fixed typo 'MP1' -> 'LP1', updated code formatting

File: ports/stm32/machine_uart.c [hunk 2]
```diff
@@ -76,7 +76,15 @@
 STATIC void pyb_uart_print(const mp_print_t *print, mp_obj_t self_in, mp_print_kind_t kind) {
     pyb_uart_obj_t *self = MP_OBJ_TO_PTR(self_in);
     if (!self->is_enabled) {
-        mp_printf(print, "UART(%u)", self->uart_id);
+        #ifdef LPUART1
+        if (self->uart_id == PYB_LPUART_1) {
+            mp_printf(print, "UART('LP1')");
+        } else {
+        #endif
```
Comment (dpgeorge) [inline ports/stm32/machine_uart.c:83]: simpler to use the following pattern:
```c
#ifdef LPUART1
if (...) {
} else
#endif
{
    mp_printf(...);
}
```

File: ports/stm32/mpconfigboard_common.h [hunk 3]
```diff
@@ -217,7 +217,7 @@
 #define PYB_EXTI_NUM_VECTORS (24)
 #define MICROPY_HW_MAX_I2C (4)
 #define MICROPY_HW_MAX_TIMER (17)
-#define MICROPY_HW_MAX_UART (8)
+#define MICROPY_HW_MAX_UART (9)
```
Comment (dpgeorge) [inline ports/stm32/mpconfigboard_common.h:220]: might be clearer to do something like this:
```c
#define MICROPY_HW_MAX_LPUART (1)
#define MICROPY_HW_MAX_UART (8)
```
(ie last line there unchanged), and then in `mpconfigport.h`:
```c
struct _pyb_uart_obj_t *pyb_uart_obj_all[MICROPY_HW_MAX_UART + MICROPY_HW_MAX_LPUART]; \
```

File: ports/stm32/uart.h [hunk 4]
```diff
@@ -31,14 +31,26 @@
...
+    #else
     PYB_UART_6 = 6,
+    #endif
     PYB_UART_7 = 7,
     PYB_UART_8 = 8,
+    #if defined(STM32H7)
+    PYB_LPUART_1 = 9,
+    #else
     PYB_UART_9 = 9,
+    #endif
```
Comment (dpgeorge) [inline ports/stm32/uart.h:53]: all this could be simplified to:
```c
#ifdef LPUART1
PYB_LPUART_1 = MICROPY_HW_MAX_UART,
#endif
```

**Comment by chrismas9** [comment]
Thanks for the review and suggestions. Once this is cleaned up I can add LPUART1 to `mpconfigboard.h` for some boards and update the docs.

**Comment by dpgeorge** [comment]
@chrismas9 I'd like to merge this before the STM32G4 addition.  If you can't get to updating the PR just let me know and I can do it.

**Comment by H-Grobben** [comment]
I've added this into the STM32G4 (#6911), but it would also apply to other families of the STM32, just didn't test it, as I don't have the hardware to test it. I did build all STM32 families (ci script nucleo and pyb boards)

**Comment by dpgeorge** [comment]
Thanks for updating the PR.  Merged in 9d674cf7ab8570637dfab23811800e3b98babad9 with some minor changes:
- fixed so it raises an exception creating an LP1 UART (without specifying a baudrate or any other params) if the TX/RX pins are not defined
- reverted stray changes to some config values in mpconfigboard_common.h
- made AF_FN_LPUART distinct from AF_FN_UART to avoid changing `mp_hal_pin_config_alt`

I also enabled LPUART1 on NUCLEO_WB55 in a separate commit, 03a64f20771cd0582af61de31811294456fa695e

**Pull request closed without merging**